exposed-dao/src/main/kotlin/org/jetbrains/exposed/dao/EntityClass.kt

```diff
@@ -954,12 +956,79 @@ abstract class EntityClass<ID : Comparable<ID>, out T : Entity<ID>>(
+        forUpdate: Boolean? = null
+    ): List<T> {
+        val parentTable = refColumns.values.firstOrNull()?.table as? CompositeIdTable
+        requireNotNull(parentTable) { "RefColumns should have reference to CompositeIdTable" }
+        if (references.isEmpty()) return emptyList()
+        val distinctRefIds = references.distinct().map { EntityID(it, parentTable) }
+        val transaction = TransactionManager.current()
+        val cache = transaction.entityCache
+        val keepLoadedReferenceOutOfTransaction = transaction.db.config.keepLoadedReferencesOutOfTransaction
+        if (refColumns.keys.all { it.columnType is EntityIDColumnType<*> }) {
+            val toLoad = distinctRefIds.filter {
+                cache.referrers[delegateRefColumn]?.containsKey(it)?.not() ?: true
+            }
+            if (toLoad.isNotEmpty()) {
+                val findQuery = find { refColumns.keys.toList() inList toLoad.map { it.value } }
+                val entities = getEntities(forUpdate, findQuery)
+                val result = entities.groupByReference<CompositeID>(refColumns = refColumns)
+
+                distinctRefIds.forEach { id ->
+                    cache.getOrPutReferrers(id, delegateRefColumn) { result[id.value]?.let { SizedCollection(it) } ?: emptySized() }.also {
+                        if (keepLoadedReferenceOutOfTransaction) {
+                            cache.find(this, id as EntityID<ID>)?.storeReferenceInCache(delegateRefColumn, it)
+                        }
+                    }
+                }
+            }
+
+            return distinctRefIds.flatMap { cache.getReferrers<T>(it, delegateRefColumn)?.toList().orEmpty() }
+        } else {
+            val baseQuery = searchQuery(Op.build { refColumns.keys.toList() inList distinctRefIds.map { it.value } })
+            val findQuery = wrapRows(baseQuery)
+            val entities = getEntities(forUpdate, findQuery).distinct()
+            val result = entities.groupByReference<CompositeID>(refColumns = refColumns)
+
+            result.forEach { (id, values) ->
+                val parentEntityId: EntityID<*> = parentTable.selectAll().where { parentTable.id eq id }.single()[parentTable.id]
+
+                cache.getOrPutReferrers(parentEntityId, delegateRefColumn) { SizedCollection(values) }.also {
+                    if (keepLoadedReferenceOutOfTransaction) {
+                        val childEntity = find { refColumns.keys.toList() inList listOf(id) }.firstOrNull()
+                        childEntity?.storeReferenceInCache(delegateRefColumn, it)
+                    }
+                }
+            }
+            return entities
+        }
+    }
+
     private fun getEntities(forUpdate: Boolean?, findQuery: SizedIterable<T>): List<T> = when (forUpdate) {
         true -> findQuery.forUpdate()
         false -> findQuery.notForUpdate()
         else -> findQuery
     }.toList()
 
+    private fun <R> List<T>.groupByReference(
```

Comment:
It's always used with one argument, so it could be split into 2 independen functions.